react/src/Calendar/Calendar.tsx

```diff
@@ -31,32 +31,67 @@ export interface CalendarProps
    * @param {Date | null} date The new selected date.
    */
   onChange?: (date: Date | null) => void
+  /**
+   * Callback fired when the currently viewed month changes.
+   * @param {number} currMonth The new month being viewed.
+   */
+  onMonthChange?: (currMonth: number) => void
+  /**
+   * Callback fired when the currently viewed year changes.
+   * @param {number} currYear The new year being viewed.
+   */
+  onYearChange?: (currYear: number) => void
   /** The default selected date, used if input is uncontrolled */
   defaultValue?: Date | null
   /** Function to determine whether a date should be made unavailable. */
   isDateUnavailable?: (d: Date) => boolean
+  /** Whether to render a loading state. */
+  isLoading?: boolean
   /** Color scheme for component */
   colorScheme?: ThemingProps<'Calendar'>['colorScheme']
 }
 
 export const Calendar = forwardRef<CalendarProps, 'input'>(
   (
-    { value, onChange, defaultValue, showTodayButton = true, ...props },
+    {
+      value,
+      onChange,
+      onMonthChange,
+      onYearChange,
+      defaultValue,
+      showTodayButton = true,
+      ...props
+    },
     initialFocusRef,
   ) => {
     const styles = useMultiStyleConfig('Calendar', props)
-
     const [internalValue, setInternalValue] = useControllableState({
       value,
       onChange,
-      defaultValue,
+      defaultValue: defaultValue,
+    })
+
+    const [currMonth, setCurrMonth] = useControllableState<number>({
+      defaultValue:
+        props.defaultFocusedDate?.getMonth() ?? new Date().getMonth(),
+      onChange: onMonthChange,
+    })
+
+    const [currYear, setCurrYear] = useControllableState<number>({
+      defaultValue:
+        props.defaultFocusedDate?.getFullYear() ?? new Date().getFullYear(),
+      onChange: onYearChange,
     })
```

Comment:
should this be moved into the provider instead? So the logic all lives in the provider, and RangeCalendar will not need duplicated code too. Maybe `internalValue` can move in too haha, not sure why I extracted it (but maybe you'll re-figure it out)

Reply:
done as recommended. 

I'm guessing that you kept `internalValue` outside the context so that you could put the 20 lines of array manipulation inside `RangeCalendar.tsx` while the simple `Calendar.tsx` uses simple React state. I actually like this separation (so all range-related logic is in the 1 file), so I've left it outside.